### DISCUSSION TITLE
IPSEC NAT66 with NAT-T detected

### DISCUSSION THREAD

**Discussion by ashot-a** [strongswan/strongswan#1603] [Q&A] [answered]
**System (please complete the following information):**
 - OS: [Ubuntu 18.04]
 - Kernel version (if applicable): [Linux 5.10.0-051000-generic, x86_64]
 - strongSwan version(s): [strongSwan 5.6.2]
 - Tested/confirmed with the latest version: [yes/no]

**Describe the bug**
Configured route base VPN for IP6 on IKEv2 with NAT-T involved. I see tunnel is UP on both sides but traffic is not 
forwarded on strong swan side. It receives ESP UDP-ENCAP ESP packets on eth1 but doesn’t forward to eth2
The sysctl net.ipv6.conf.all.forwarding=1 and not ip6tables rules are configured

**To Reproduce**
1. Configure IP6 site to site VPN NAT involved in between
2 conn def-tun
      fragmentation=yes
      dpddelay=1
      dpdtimeout=3
      dpdaction=restart
      #ike=aes128-sha1-modp1024,aes256-sha1-modp1024
      #esp=aes128-sha1-modp1024,aes256-sha1-modp1024
      authby=secret
      type=tunnel
conn Nat_IPv6
           also=def-tun
           left=fd00:bbbb:1:1::1
           #leftid="fd00:bbbb:1:1::1"
           leftsubnet=fd00:cccc:1:3::/64
           #leftfirewall=yes
           right=fd00:1:1:1::22
           rightid="suresh@vmware.com"
           rightfirewall=yes
           ike= aes256-sha256-modp3072
           esp= aes256-sha256-modp3072
           ikelifetime=86400s
           keylife=28800s
           rightsubnet=fd00:1:1::/64
           #keyingtries=1
           keyexchange=ikev2
           auto=route
3.  check ipsec status 
root@SU-5SI-nvs-gw1:~# ipsec statusall
Status of IKE charon daemon (strongSwan 5.6.2, Linux 5.10.0-051000-generic, x86_64):
  uptime: 9 seconds, since Mar 15 16:37:30 2023
  malloc: sbrk 2564096, mmap 0, used 653088, free 1911008
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 8
  loaded plugins: charon aesni aes rc2 sha2 sha1 md4 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm attr kernel-netlink resolve socket-default connmark stroke updown eap-mschapv2 xauth-generic counters
Listening IP addresses:
  192.168.0.140
  169.254.6.26
  fd00:bbbb:1:1::1
  169.254.11.26
  fd00:cccc:1:1::1
  172.17.0.1
Connections:
    Nat_IPv6:  fd00:bbbb:1:1::1...fd00:1:1:1::22  IKEv2, dpddelay=1s
    Nat_IPv6:   local:  [fd00:bbbb:1:1::1] uses pre-shared key authentication
    Nat_IPv6:   remote: [suresh@vmware.com] uses pre-shared key authentication
    Nat_IPv6:   child:  fd00:cccc:1:3::/64 === fd00:1:1::/64 TUNNEL, dpdaction=restart
Routed Connections:
    Nat_IPv6{1}:  ROUTED, TUNNEL, reqid 1
    Nat_IPv6{1}:   fd00:cccc:1:3::/64 === fd00:1:1::/64
Security Associations (1 up, 0 connecting):
    Nat_IPv6[1]: ESTABLISHED 7 seconds ago, fd00:bbbb:1:1::1[fd00:bbbb:1:1::1]...fd00:1:1:1::22[suresh@vmware.com]
    Nat_IPv6[1]: IKEv2 SPIs: 00d0c52e4e56dd8d_i 590b55915d083cef_r*, pre-shared key reauthentication in 23 hours
    Nat_IPv6[1]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_3072
    Nat_IPv6{2}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: ceb578ef_i 4ea314f2_o
    Nat_IPv6{2}:  AES_CBC_256/HMAC_SHA2_256_128, 0 bytes_i, 0 bytes_o, rekeying in 7 hours
    Nat_IPv6{2}:   fd00:cccc:1:3::/64 === fd00:1:1::/64
5. ...
6. ...

**Expected behavior**

We do expect traffic to be forwarded from eth1 interface to eth2 .

**Logs/Backtraces**
If applicable, add logs or backtraces to help explain your problem.
Initiating ping from client behind right side VPN to left side client behind strong swan gateway.
We see UDP-encap ESP packets received on eth1 but doesn't forward to eth2

tcpdump -ni eth1

6:43:23.606961 IP6 fd00:1:1:1::22.4500 > fd00:bbbb:1:1::1.4500: UDP-encap: ESP(spi=0xceb578ef,seq=0xd12), length 152
16:43:23.711000 IP6 fd00:1:1:1::22.4500 > fd00:bbbb:1:1::1.4500: UDP-encap: ESP(spi=0xceb578ef,seq=0xd13), length 152
16:43:23.814939 IP6 fd00:1:1:1::22.4500 > fd00:bbbb:1:1::1.4500: UDP-encap: ESP(spi=0xceb578ef,seq=0xd14), length 152
16:43:23.919037 IP6 fd00:1:1:1::22.4500 > fd00:bbbb:1:1::1.4500: UDP-encap: ESP(spi=0xceb578ef,seq=0xd15), length 152
16:43:24.023116 IP6 fd00:1:1:1::22.4500 > fd00:bbbb:1:1::1.4500: UDP-encap: ESP(spi=0xceb578ef,seq=0xd16), length 152

**Additional context**
The "ipsec statusall " output
root@SU-5SI-nvs-gw1:~# ipsec statusall
Status of IKE charon daemon (strongSwan 5.6.2, Linux 5.10.0-051000-generic, x86_64):
  uptime: 9 seconds, since Mar 15 16:37:30 2023
  malloc: sbrk 2564096, mmap 0, used 653088, free 1911008
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 8
  loaded plugins: charon aesni aes rc2 sha2 sha1 md4 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp agent xcbc hmac gcm attr kernel-netlink resolve socket-default connmark stroke updown eap-mschapv2 xauth-generic counters
Listening IP addresses:
  192.168.0.140
  169.254.6.26
  fd00:bbbb:1:1::1
  169.254.11.26
  fd00:cccc:1:1::1
  172.17.0.1
Connections:
    Nat_IPv6:  fd00:bbbb:1:1::1...fd00:1:1:1::22  IKEv2, dpddelay=1s
    Nat_IPv6:   local:  [fd00:bbbb:1:1::1] uses pre-shared key authentication
    Nat_IPv6:   remote: [suresh@vmware.com] uses pre-shared key authentication
    Nat_IPv6:   child:  fd00:cccc:1:3::/64 === fd00:1:1::/64 TUNNEL, dpdaction=restart
Routed Connections:
    Nat_IPv6{1}:  ROUTED, TUNNEL, reqid 1
    Nat_IPv6{1}:   fd00:cccc:1:3::/64 === fd00:1:1::/64
Security Associations (1 up, 0 connecting):
    Nat_IPv6[1]: ESTABLISHED 7 seconds ago, fd00:bbbb:1:1::1[fd00:bbbb:1:1::1]...fd00:1:1:1::22[suresh@vmware.com]
    Nat_IPv6[1]: IKEv2 SPIs: 00d0c52e4e56dd8d_i 590b55915d083cef_r*, pre-shared key reauthentication in 23 hours
    Nat_IPv6[1]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_3072
    Nat_IPv6{2}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: ceb578ef_i 4ea314f2_o
    Nat_IPv6{2}:  AES_CBC_256/HMAC_SHA2_256_128, 0 bytes_i, 0 bytes_o, rekeying in 7 hours
    Nat_IPv6{2}:   fd00:cccc:1:3::/64 === fd00:1:1::/64

**Comment by ashot-a** [comment]
Can you please help if provided versions support NAT66 and if we do have right Strongswan config to make it work?
Thanks Ashot

**Comment by tobiasbrunner** [comment, marked as answer]
> Can you please help if provided versions support NAT66

Support in what way?

> if we do have right Strongswan config to make it work?

To make what work exactly?

Your description is seriously lacking information (see #196).

**Reply by ashot-a** [reply]
From provided Strongswan config you can see that traffic is destined  to IP6 client behind Strongswan gateway. When there is no NAT-T detected at IKEv2 tunnel establishment, then tunnel traffic reaches to IP6 client. When there is one to one IP6 NAT applied in between  we see that traffic doesn't forwarded from eth1 to eth2 on Srongswan gateway.  
Answering to your  "To make what work exactly?"  question.  I meant IP6 site to site VPN with NAT-T IP6 detected.  In other words there is IP6 NAT applied between Strongswan gateway and our edge.

**Reply by tobiasbrunner** [reply]
I don't see how the UDP encapsulation of the ESP packets should affect forwarding **after** decryption. Note that the traffic counters in the status output don't show any processed packages (i.e. there was nothing to forward).

**Reply by ashot-a** [reply]
I noticed that too, For that reason I would like to clarify about compatibility between Ubuntu 18.04 with   Linux 5.10.0-051000-generic, x86_64 kernel and Strongswan 5.6.2 version from point of view NAT66 case. Do you know if all these versions are compatible each others in terms for NAT66 case. Thanks

**Reply by tobiasbrunner** [reply]
No idea. But apparently the SA was installed, so the kernel seems to support UDP-encap for IPv6 (it was originally added with 5.8). Not sure if there have been any significant bugs that might not have been backported to 5.10. Anyway, what happens beyond that you need to debug via traffic/error counters, maybe captures and possibly netfilter tracing.
